### PR TITLE
4th PR of Python/GT4Py implementation (Shallow Water Driver & Visualization)

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

This is the fourth of ~5 PRs that will introduce a Python/GT4Py implementation of the Shallow Water code.

This PR introduces: 

- The Python implementation of the Shallow Water Driver code 
- Visualization using matplotlib.
- Add a shallow_water.yml config file to be used in visualization. 
- Add runtime section in test_shallow_water_config.yml to use a smaller amount of run_steps for testing driver code. 
- Update .gitignore to not include .gif files output by visualization. 
- Add env variables to [.github/workflows/macos_intel.yml](https://github.com/NOAA-GSL/SENA-shallow-water/compare/feature/python-state?expand=1#diff-615c12e4300e83697176d001cde7564689560b65caa79aeefa15e21455122b41) workflow to attempt to point C++ compiler to Intel icpc instead of clang for serialbox install.

To test the driver code (config file as well as restart file):
```
cd src/python
mpiexec -n <number of processors> python -m shallow_water ../../parm/shallow_water.yml
mpiexec -n <number of processors> python -m shallow_water ../../parm/shallow_water.yml --filename "<netCDF restart filepath>"
```
To test the visualization code:
`mpiexec -n 1 python -m shallow_water_visualization ../../parm/shallow_water.yml`

**No unit tests (Pytests) were added for these two scripts.** 

Tested on laptop with 1,2,4, & 9 processes:
`mpiexec -n 9 --oversubscribe python -m pytest --with-mpi -s --disable-pytest-warnings`

And for non-mpi tests:
`mpiexec -n 1 python -m pytest`

## Types of changes

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [x] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-shallow-water/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if appropriate)

## Reviewers' checklist

The following is a checklist for reviewers to consider when reviewing this pull request.

#### Implementation

- [ ] Does this code do what it is supposed to do?
- [ ] Can the code be simplified?
- [ ] Does the code add unwanted compile-time or run-time dependencies?
- [ ] Is the code at the correct abstraction level?
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If so, why wasn't it used?
- [ ] Can you think of a different solution that is substantially better in terms of code maintenance, readability, or performance?

#### Logic Errors and Bugs

- [ ] Can you think of a use case in which the code does not behave properly?
- [ ] Can you think of any inputs or external events that could break the code?

#### Error handling and logging

- [ ] Is error handling done correctly?
- [ ] Should any logging or debuggging be added or removed?
- [ ] Are error messages user friendly?

#### Usability

- [ ] Is the API intuitive to use?
- [ ] Is the API well documented?

#### Testing and testability

- [ ] Do the tests pass?
- [ ] Is the code testable?
- [ ] Do the tests reasonably cover the code change?
- [ ] Are there additional edge cases or inputs that should be tested?

#### Performance

- [ ] Do you think the code change will negatively impact performance?
- [ ] Do you see any potential to improve the performance of the code?

#### Dependencies

- [ ] Does this change require updates to documenation, configuration?  If so, what that done?
- [ ] Does this change affect backward compatibility?

#### Readability

- [ ] Is the code easy to understand?
- [ ] Can the readability be improved with smaller routines?
- [ ] Can the readability be improved with different subroutine/variable names?
- [ ] Is the data flow understandable?
- [ ] Could comments in the code be improved to add clarity?
- [ ] Is there any commented out code?
- [ ] Does the code conform to the style guide (Fortran style compliance cannot be checked automatically)

### CODE REVIEW

File: src/python/shallow_water.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        for i in range(g.xps, g.xpe + 1):
+            for j in range(g.yps, g.ype + 1):
+                dsqr = ((i-1) * g.dx - xmid)**2 + ((j-1) * g.dy - ymid)**2
+                h[i - g.xps,j - g.yps] = 5000.0 + np.exp(-dsqr / sigma**2) * (mc.h0 - 5000.0)
+        s = ShallowWaterState(g, gtc, h=h)
+
+
+    # Write out the initial state if needed
+    if (runtime['output_interval_steps'] <= runtime['run_steps'] and filename is None):
+        s.write(f"../../test/test_output/swout_{round(s.clock / m.dt):07d}.nc")
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water.py]: This program is not a test so the output should not go to `../../test/test_output`.  It should go to the current directory. We could add an `output_directory` option in the runtime part of the YAML to specify where it should go.  But absent that, it needs to go to the current directory.

File: src/python/shallow_water.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    # Run the model
+    for t in range(0, runtime['run_steps'], runtime['output_interval_steps']):
+        m.adv_nsteps(s, min(runtime['output_interval_steps'], runtime['run_steps'] - t))
+        if (runtime['output_interval_steps']):
+            s.write(f"../../test/test_output/swout_{round(s.clock / m.dt):07d}.nc")
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water.py]: Same as above.

File: .github/workflows/macos_intel.yml [hunk 3]
```diff
@@ -174,6 +174,12 @@ jobs:
           cd ${HOME}/serialbox
           mkdir build
           cd build
+          XCODE_PATH=$(xcode-select -p)
+          XCODE_HEADERS=${XCODE_PATH}/SDKs/MacOSX.sdk/usr/include
+          XCODE_FRAMEWORK=${XCODE_PATH}/SDKs/MacOSX.sdk/System/Library/Frameworks
+          export C_INCLUDE_PATH+=:${XCODE_PATH}/SDKs/MacOSX.sdk/usr/include
+          export CPLUS_INCLUDE_PATH+=:${XCODE_PATH}/SDKs/MacOSX.sdk/usr/include
+          export LIBRARY_PATH+=:${XCODE_PATH}/SDKs/MacOSX.sdk/usr/lib
```
Comment (christopherwharrop-noaa) [inline .github/workflows/macos_intel.yml]: All changes in this file should be removed as the other PR applies the fix.

File: src/python/shallow_water_visualization.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+## Plot animation
+X, Y = np.meshgrid(np.linspace(0, g.xmax, g.npx), np.linspace(0, g.ymax, g.npy))
+fig, axs = plt.subplots()
+norm = ColormapNormalize(vmin=4990, vmax=5030, v1=4995, v2=5005)
+pc = axs.pcolormesh(X, Y, s.h.data, cmap='nipy_spectral', norm=norm)
+cb = fig.colorbar(pc, ax=axs, extend='both')
+cb.set_ticks([4990, 4995, 4996, 4997, 4998, 4999, 5000, 5001, 5002, 5003, 5004, 5005, 5030])
+anim = animation.FuncAnimation(fig, animate, interval=125, frames=runtime['run_steps'])
+anim.save('../../test/test_output/test.gif')
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water_visualization.py]: Again.  This is not a test program in the test suite.  It's output should go to the current directory.

File: src/python/shallow_water_visualization.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                                            left=-np.inf, right=np.inf))
+
+    def inverse(self, value):
+        y, x = [self.vmin, self.v1, self.v2, self.vmax], [0, 0.1, 0.9, 1]
+
+        return np.interp(value, x, y, left=-np.inf, right=np.inf)
+
+comm = MPI.COMM_WORLD
+
+config_file = "../../parm/shallow_water.yml"
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water_visualization.py]: You may want to generalize this in the same way you did with the main program by using command line arguments for consistency.

File: test/test_input/test_shallow_water_config.yml [hunk 6]
```diff
@@ -11,6 +11,12 @@ model:
   b0: 4.0
   h0: 5.0
 
+runtime:
```
Comment (christopherwharrop-noaa) [inline test/test_input/test_shallow_water_config.yml]: None of the tests use this section of the namelist.  So this won't do anything.  The main program and visualization will fail with this YAML file because the timestep is not valid.  This was not intended for testing the full integration, but rather just the configuration classes.

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
There are a few issues that need resolving but nothing too serious.  You should probably update the text of the description to indicate that testing these should be done with:

```
mpiexec -n N python -m shallow_water ../../parm/shallow_water.yml
```

Where N is the number of processes you want to use.

The visualization can only be run on one process I think.

```
mpiexec -n 1 python -m shallow_water_visualization ../../parm/shallow_water.yml
```

File: src/python/shallow_water.py (hunk 1 above)
Comment (robgpita) [reply src/python/shallow_water.py]: You're right, this was a holdover from when I was testing the script, and I didn't switch it back. I removed `../../test/test_output/` from here and below, and tested it locally.

File: .github/workflows/macos_intel.yml (hunk 3 above)
Comment (robgpita) [reply .github/workflows/macos_intel.yml]: Thats for making the correct fix in PR #17 .

File: src/python/shallow_water_visualization.py (hunk 5 above)
Comment (robgpita) [reply src/python/shallow_water_visualization.py]: I restructured this script to take in command line arguments similar to `shallow_water.py`.

**Review by christopherwharrop-noaa — APPROVED** [review]
Thank you for the updates and the great enhancements to the original.
